// honkit/lib/cli/options.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = __importDefault(require("../utils/logger"));
const logOptions = {
    name: "log",
    description: "Minimum log level to display",
    values: logger_1.default.LEVELS.keySeq()
        .map((s) => {
        return s.toLowerCase();
    })
        .toJS(),
    defaults: "info",
};
const formatOption = {
    name: "format",
    description: "Format to build to",
    values: ["website", "json", "ebook"],
    defaults: "website",
};
const timingOption = {
    name: "timing",
    description: "Print timing debug information",
    defaults: false,
};
const reloadOption = {
    name: "reload",
    description: "Prune cache. Remove file cache",
    defaults: false,
};
exports.default = {
    log: logOptions,
    format: formatOption,
    timing: timingOption,
    reload: reloadOption,
};
