// honkit/lib/constants/defaultBlocks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const immutable_1 = __importDefault(require("immutable"));
const templateBlock_1 = __importDefault(require("../models/templateBlock"));
exports.default = immutable_1.default.Map({
    html: new templateBlock_1.default({
        name: "html",
        process: function (blk) {
            return blk;
        },
    }),
    code: new templateBlock_1.default({
        name: "code",
        process: function (blk) {
            return {
                html: false,
                body: blk.body,
            };
        },
    }),
    markdown: new templateBlock_1.default({
        name: "markdown",
        process: function (blk) {
            return this.book.renderInline("markdown", blk.body).then((out) => {
                return { body: out };
            });
        },
    }),
    asciidoc: new templateBlock_1.default({
        name: "asciidoc",
        process: function (blk) {
            return this.book.renderInline("asciidoc", blk.body).then((out) => {
                return { body: out };
            });
        },
    }),
    markup: new templateBlock_1.default({
        name: "markup",
        process: function (blk) {
            return this.book.renderInline(this.ctx.file.type, blk.body).then((out) => {
                return { body: out };
            });
        },
    }),
});
